pkg/skaffold/build/gcb/status.go

```diff
@@ -0,0 +1,272 @@
+/*
+Copyright 2021 The Skaffold Authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package gcb
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"strings"
+	"sync"
+	"time"
+
+	"google.golang.org/api/cloudbuild/v1"
+	"k8s.io/apimachinery/pkg/util/wait"
+
+	"github.com/GoogleContainerTools/skaffold/pkg/skaffold/gcp"
+)
+
+var (
+	// maintain single instance of `statusReporter` per skaffold process
+	reporter     *statusManagerImpl
+	reporterOnce sync.Once
+	// maintain single instance of the GCB client per skaffold process
+	client     *cloudbuild.Service
+	clientOnce sync.Once
+)
+
+// statusManager provides an interface for getting the status of GCB jobs.
+// It reduces the number of requests made to the GCB API by using a single `List` call for
+// all concurrently running builds for a project instead of multiple per-job `Get` calls.
+// The statuses are polled with an independent exponential backoff strategy per project until success, cancellation or failure.
+// The backoff duration for each `projectID` status query is reset with every new build request.
+type statusManager interface {
+	getStatus(ctx context.Context, projectID string, buildID string) <-chan result
+}
+
+// statusManagerImpl implements the `statusManager` interface
+type statusManagerImpl struct {
+	results     map[jobID]chan result
+	resultMutex sync.RWMutex
+	requests    chan jobRequest
+}
+
+// jobID represents a single build job
+type jobID struct {
+	projectID string
+	buildID   string
+}
```

Comment:
WDYT of creating types for `buildID` and `projectID` for use in the signatures and get some type checking into place:
```
type projectID string
type buildID string
```

I think the poller would benefit.